Ensure Microsoft.NETCore.App is an implicit dependency for publish.

peterhuene opened the pull request:
This commit ensures that we exclude files from the `Microsoft.NETCore.App`
platform library on publish when the platform library is
`Microsoft.AspNetCore.App`, which doesn't have an explicit dependency on
`Microsoft.NETCore.App`.

When publishing a 2.x ASP.NET application, the `MicrosoftNETPlatformLibrary`
property gets changed by ASP.NET to `Microsoft.AspNetCore.App`.  This causes
the task responsible for determining what files to copy locally to treat files
from `Microsoft.NETCore.App` and its dependencies as not being part of the
platform.

However, when publishing the application as framework-dependent, almost all the
platform files get excluded thanks to conflict resolution which has the files
as part of the platform manifest and preferred packages.  Unfortunately, the
apphost, hostpolicy, and hostfxr files are not part of these lists and as a
result will be copied locally.

This breaks framework-dependent apphost activation because hostfxr and
hostpolicy are copied locally.  Additionally, these files end up in the
deps.json file which also prevents roll-forward activation if the two files are
manually deleted.

The fix is to treat `Microsoft.NETCore.App` as an implicit dependency of
`Microsoft.AspNetCore.App` where we calculate the list of excluded packages for
a framework-dependent publish.

Fixes dotnet/cli#10602.

nguerrera commented:
I did not review in detail, but can we not consider that there is an implicit dependency from Microsoft.AspNetCore.App to Microsoft.NETCore.App?

I don't understand why this requires hard-coding of file versions, etc.

peterhuene commented:
The hard coding is effectively what should be present in the platform manifest.

I'm not happy with this solution either.  Let me investigate what it would look like to hardcode an implicit dependency to `Microsoft.NETCore.App` when resolving copy local files for publish.

peterhuene commented:
Actually, with all the recent changes to asset resolution in 3.0 (namely how we resolve the apphost), this might not need fixing any more in 3.0.  If that's the case, an implicit reference would be ideal vs. needlessly trying to replicate what I envisioned the 3.0 fix would be.

peterhuene commented:
Ok, I think making an implicit dependency on `Microsoft.NETCore.App` for exclusions is the approach we'll take for this fix.  I'll redo the PR shortly.

Commit:
Ensure Microsoft.NETCore.App is an implicit dependency for publish.

This commit ensures that we exclude files from the `Microsoft.NETCore.App`
platform library on publish when the platform library is
`Microsoft.AspNetCore.App`, which doesn't have an explicit dependency on
`Microsoft.NETCore.App`.

When publishing a 2.x ASP.NET application, the `MicrosoftNETPlatformLibrary`
property gets changed by ASP.NET to `Microsoft.AspNetCore.App`.  This causes
the task responsible for determining what files to copy locally to treat files
from `Microsoft.NETCore.App` and its dependencies as not being part of the
platform.

However, when publishing the application as framework-dependent, almost all the
platform files get excluded thanks to conflict resolution which has the files
as part of the platform manifest and preferred packages.  Unfortunately, the
apphost, hostpolicy, and hostfxr files are not part of these lists and as a
result will be copied locally.

This breaks framework-dependent apphost activation because hostfxr and
hostpolicy are copied locally.  Additionally, these files end up in the
deps.json file which also prevents roll-forward activation if the two files are
manually deleted.

The fix is to treat `Microsoft.NETCore.App` as an implicit dependency of
`Microsoft.AspNetCore.App` where we calculate the list of excluded packages for
a framework-dependent publish.

Fixes dotnet/cli#10602.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs:
```diff
@@ -12,6 +12,9 @@ namespace Microsoft.NET.Build.Tasks
     internal class ProjectContext
     {
+        private const string NetCorePlatformLibrary = "Microsoft.NETCore.App";
+        private const string AspNetPlatformLibrary = "Microsoft.AspNetCore.App";
```
I think this can also be Microsoft.AspNetCore.*All* on 2.x.

nguerrera replied:
Maybe consider any platform library that is not base Microsoft.NETCore.App to depend on Microsoft.NETCore.App?

peterhuene replied:
[You are correct](https://github.com/aspnet/Universe/blob/da48de7d7580ce316784dd3e250786ffa7017c25/src/Packages/Microsoft.AspNetCore.All/build/netcoreapp3.0/Microsoft.AspNetCore.All.targets#L10).

Perhaps a check for "not empty and not `Microsoft.NETCore.App`" would be sufficient?

nguerrera replied:
Yes, that sounds good.

peterhuene replied:
Pushed.

Commit:
Treat `Microsoft.NETCore.App` as implicit dependency for all platform lib names.

This commit treats `Microsoft.NETCore.App` as an implicit dependency for any
platform library name, allowing the fix to work for both `Microsoft.AspNet.App`
and `Microsoft.AspNet.All`.

livarcocc commented:
@peterhuene let me know when this is ready for shiproom.

peterhuene commented:
@livarcocc I think it's ready.  I'll try to be available for the next shiproom to help defend, but I may have to do so from my TEALS class.

livarcocc commented:
@peterhuene don't worry. I will talk to you tomorrow morning and get the information I need. Don't worry about it when you are in class.

peterhuene commented:
This has been approved but is waiting for the branch to open for the appropriate release before merging.

peterhuene commented:
As the 2.2.1xx branch is now open for the next release, I'm merging this.
